crates/common/src/transaction.rs

```diff
@@ -207,3 +277,978 @@ pub enum DataAvailabilityMode {
+
+        let hash = self
+            .data_hashes
+            .iter()
+            .fold(hasher, |hasher, &data| hasher.chain(data.into()))
+            .finish();
+
+        TransactionHash(hash.into())
+    }
+
+    fn pack_data_pack_data_availability(&self) -> u64 {
+        let nonce = u64::from(self.nonce_data_availability_mode) << 32;
+        let fee = u64::from(self.fee_data_availability_mode);
+
+        nonce + fee
+    }
+
+    fn hash_paymaster_data(&self) -> Felt {
+        self.paymaster_data
+            .iter()
+            .fold(PoseidonHasher::default(), |hasher, data| {
+                hasher.chain(data.0.into())
+            })
+            .finish()
+            .into()
+    }
+
+    fn hash_fee_fields(&self) -> Felt {
+        PoseidonHasher::default()
+            .chain(self.tip.0.into())
+            .chain(Self::pack_gas_bound(b"L1_GAS", &self.resource_bounds.l1_gas).into())
+            .chain(Self::pack_gas_bound(b"L2_GAS", &self.resource_bounds.l2_gas).into())
+            .finish()
+            .into()
+    }
+
+    fn pack_gas_bound(name: &[u8], bound: &ResourceBound) -> Felt {
+        let mut buffer: [u8; 32] = Default::default();
+        let (remainder, max_price) = buffer.split_at_mut(128 / 8);
+        let (gas_kind, max_amount) = remainder.split_at_mut(64 / 8);
+
+        let padding = gas_kind.len() - name.len();
+        gas_kind[padding..].copy_from_slice(name);
+        max_amount.copy_from_slice(&bound.max_amount.0.to_be_bytes());
+        max_price.copy_from_slice(&bound.max_price_per_unit.0.to_be_bytes());
+
+        Felt::from_be_bytes(buffer).expect("Packed resource should fit into felt")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::macro_prelude::*;
+
+    #[rstest::rstest]
+    #[test]
+    #[case::declare_v0(declare_v0(), ChainId::GOERLI_INTEGRATION)]
+    #[case::declare_v1(declare_v1(), ChainId::SEPOLIA_TESTNET)]
+    #[case::declare_v2(declare_v2(), ChainId::SEPOLIA_TESTNET)]
+    #[case::declare_v3(declare_v3(), ChainId::GOERLI_INTEGRATION)]
+    #[case::deploy(deploy(), ChainId::GOERLI_TESTNET)]
+    #[case::deploy_legacy(deploy_legacy(), ChainId::GOERLI_TESTNET)]
+    #[case::deploy_account_v1(deploy_account_v1(), ChainId::MAINNET)]
+    #[case::deploy_account_v3(deploy_account_v3(), ChainId::GOERLI_INTEGRATION)]
+    #[case::invoke_v0(invoke_v0(), ChainId::GOERLI_TESTNET)]
+    #[case::invoke_v0_legacy(invoke_v0_legacy(), ChainId::GOERLI_TESTNET)]
+    #[case::invoke_v1(invoke_v1(), ChainId::MAINNET)]
+    #[case::invoke_v3(invoke_v3(), ChainId::SEPOLIA_TESTNET)]
+    #[case::l1_handler(l1_handler(), ChainId::MAINNET)]
+    #[case::l1_handler_v07(l1_handler_v07(), ChainId::MAINNET)]
+    #[case::l1_handler_legacy(l1_handler_legacy(), ChainId::GOERLI_TESTNET)]
+    fn verify_hash(#[case] transaction: Transaction, #[case] chain_id: ChainId) {
+        assert!(transaction.verify_hash(chain_id));
+    }
+
+    fn declare_v0() -> Transaction {
```

Comment:
these values are copied form fixtures - right?

Reply:
Most of them. A couple of them I pulled from the actual networks.